src/main/java/seedu/address/logic/parser/ParserUtil.java

```diff
@@ -131,46 +134,71 @@ public static Set<Tag> parseTags(Collection<String> tags) throws ParseException
+     * Parses a {@code String start} {@code String end} into an {@code Range}. Both start and
+     * end must be non-null.
      * Leading and trailing whitespaces will be trimmed.
      *
-     * @throws ParseException if the given {@code start} and {@code end} is invalid.
+     * @param start Non-null string containing the start date
+     * @param end Non-null string containing the end date
+     * @throws ParseException if the given {@code start} and {@code end} is invalid, or if
+     *      the end date is before the start date
      */
-    public static ArrayList<Date> parseDate(String start, String end) throws ParseException {
+    public static Range parseNonNullRange(String start, String end) throws ParseException {
         requireNonNull(end);
         requireNonNull(start);
-        String trimmedDateEnd = end.trim();
-        String trimmedDateStart = start.trim();
-        ArrayList<Date> validDate = new ArrayList<>();
 
         try {
-            Date dateEnd = Date.of(trimmedDateEnd);
-            Date dateStart = Date.of(trimmedDateStart);
-            if (dateEnd.isBefore(dateStart)) {
-                throw new ParseException(Date.MESSAGE_INVALID_END_DATE);
-            }
-            validDate.add(dateStart);
-            validDate.add(dateEnd);
-            return validDate;
+            String trimmedStart = start.trim();
+            String trimmedEnd = end.trim();
+
+            Date startDate = Date.of(trimmedStart);
+            Date endDate = Date.of(trimmedEnd);
+
+            return Range.createNonNullRange(startDate, endDate);
+        } catch (DateTimeParseException e) {
+            throw new ParseException(Date.MESSAGE_CONSTRAINTS);
+        } catch (EndBeforeStartException e) {
+            throw new ParseException(Range.MESSAGE_INVALID_END_DATE);
+        }
+    }
+
+    /**
+     * Parses a {@code String start} {@code String end} into an {@code Range}.
+     * Start and end can be null.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @param start String containing the start date / null
+     * @param end String containing the end date / null
+     * @throws ParseException if the given {@code start} and {@code end} is invalid, or if
+     *      the end date is before the start date
+     */
+    public static Range parseNullableRange(String start, String end) throws ParseException {
+        try {
+            boolean hasStart = start != null;
+            boolean hasEnd = end != null;
+            Date startDate = hasStart ? Date.of(start.trim()) : null;
+            Date endDate = hasEnd ? Date.of(end.trim()) : null;
+
+            return Range.createNullableRange(startDate, endDate);
         } catch (DateTimeParseException e) {
             throw new ParseException(Date.MESSAGE_CONSTRAINTS);
+        } catch (EndBeforeStartException e) {
+            throw new ParseException(Range.MESSAGE_INVALID_END_DATE);
```

Comment:
Is there any other way to handle unbounded time ranges that does not require accepting null values into the function?

Reply:
I considered using Optionals but apparently it's discouraged having Optionals as a parameter/field

Reply:
Or perhaps catch each of the null values with the hasStart and hasEnd, but each call a different factory method for initialisation?